homeassistant/auth.py

```diff
@@ -23,18 +26,35 @@
 STORAGE_VERSION = 1
 STORAGE_KEY = 'auth'
 
+CONF_MODULES = 'modules'
+
+AUTH_MODULES = Registry()
+
+AUTH_MODULE_SCHEMA = vol.Schema({
+    vol.Required(CONF_TYPE): str,
+    vol.Optional(CONF_NAME): str,
+    # Specify ID if you have two auth module for same type.
+    vol.Optional(CONF_ID): str,
+}, extra=vol.ALLOW_EXTRA)
+
 AUTH_PROVIDERS = Registry()
 
 AUTH_PROVIDER_SCHEMA = vol.Schema({
     vol.Required(CONF_TYPE): str,
     vol.Optional(CONF_NAME): str,
     # Specify ID if you have two auth providers for same type.
     vol.Optional(CONF_ID): str,
+    vol.Optional(CONF_MODULES):
```

Comment:
Modules should not be nested under an auth provider. Instead, users should be able to enable 2fa for their account.

Reply:
So the configuration should like

```yaml
# configuration.yml
homeassistant:
  auth_providers:
    - type: a_provder
    - type: b_provider
  auth_2fa_modules:
    - type: totp
    - type: hotp_sms
```

user should be able to enable one of many module in his setting page
user should be able to select use which module after auth_provider validate

That is the sequence used in my Cisco VPN client configured with okta integration.